### ISSUE TITLE
DiffLine.{new,old}_lineno is incorrect

### ISSUE THREAD

**Issue by brandonio21** [libgit2/pygit2#797]
I have found unexpected behavior while working to apply parsed diffs. The behavior for DiffLine.old_lineno and DiffLine.new_lineno is described as:

old_lineno
    Line number in old file or -1 for added line
new_lineno
    Line number in new file or -1 for deleted line

Consider the following repro:

```py
def test_diff_parse_lineno(self):                 
    patch_str = textwrap.dedent("""\              
    diff --git a/a b/a                            
    index 7f129fd..af431f2 100644                 
    --- a/a                                       
    +++ b/a                                       
    @@ -1 +1 @@                                   
    -a contents 2                                 
    +a contents                                   
    """)                                          
                                                  
    diff = pygit2.Diff.parse_diff(patch_str)      
    patches = [patch for patch in diff]           
    self.assertEqual(1, len(patches))             
                                                  
    hunks = patches[0].hunks                      
    self.assertEqual(1, len(hunks))               
                                                  
    lines = hunks[0].lines                        
    self.assertEqual(2, len(lines))               
                                                  
    self.assertNotEqual(0, lines[0].new_lineno)  # Fails                                                     
```

In my testing, I found that these two old_lineno and new_lineno always happen to be 0. This is obviously incorrect.

**Comment by jdavid** [comment]
```c
#include <stdio.h>
#include <string.h>
#include <git2.h>

const char * text = "diff --git a/a b/a\n"
                    "index 7f129fd..af431f2 100644\n"
                    "--- a/a\n"
                    "+++ b/a\n"
                    "@@ -3 +3 @@\n"
                    "-a contents 2\n"
                    "+a contents\n";

void check(int err, const char *msg)
{
    if (err < 0) { printf("Error: %s\n", msg); exit(1); }
}

void main()
{
    git_diff *diff;
    git_patch *patch;
    const git_diff_hunk *hunk; size_t n;
    const git_diff_line *line;

    printf("%s", text);
    printf("\n\n");

    check(git_diff_from_buffer(&diff, text, strlen(text)), "git_diff_from_buffer");
    check(git_patch_from_diff(&patch, diff, 0), "git_patch_from_diff");
    check(git_patch_get_hunk(&hunk, &n, patch, 0), "git_patch_get_hunk");
    for (int i=0; i<n; i++)
    {
      check(git_patch_get_line_in_hunk(&line, patch, 0, i), "git_patch_get_line_in_hunk");
      printf("%d: %d -> %d\n", i, line->old_lineno, line->new_lineno);
    }
}
```

I have the same result with this c code. If I have done a mistake, can you spot it? Otherwise this may be a libgit2 issue, then it would be better to report it in the libgit2 issue tracker.

**Comment by jdavid** [comment]
Update: Just checked with the latest version v0.27.5, and still broken. Apparently this is not going to be backported to the 0.27 branch, so we really need to wait for v0.28.0 to be released.

**Comment by pks-t** [comment]
On Sat, Oct 06, 2018 at 02:45:57AM -0700, J. David Ibáñez wrote:
> Update: Just checked with the latest version v0.27.5, and still
> broken. Apparently this is not going to be backported to the
> 0.27 branch, so we really need to wait for v0.28.0 to be
> released.

We always try hard to separate security from bugfix releases, so
that's why there was no backport of that fix. I plan to prepare a
bugfix release in the coming week and release it in the
not-too-distant future. Please feel free to leave a note in the
libgit2 PR for this bugfix to nominate it for backporting.

**Comment by jdavid** [comment]
Thanks @pks-t , it was a note to myself to not check again until the release of v0.28

Which is fine to me, but maybe @brandonio21 wishes to ask for backporting.

**Comment by jdavid** [comment]
This has been fixed in libgit2 0.27.7 , closing
